### PR TITLE
Retry service maintenance

### PR DESCRIPTION
**Pull request by nlordell**
This PR changes the service maintenance component to retry updating to the latest block after some delay instead of waiting until a new block is observed.

This is done for two reasons:
1. In order to reduce alert noise because of maintenance update delays (I made sure to keep track of success and failure so we can setup dashboards and alerts for it).
2. Because we want maintenance to run so we are at the "latest" block always. Without retrying, we stay 12 seconds knowing that there is a new block and we have stale data before trying to update again. So, conceptually, retrying here is not so much of a sin.

Furthermore, in terms of the first point, changing the parameters of the alerts is not really possible. The issue is that the current alert measures how stale our data is (i.e. how far the last updated block is from the latest block). This PR proposes actively trying to make our data less stale (retrying to update to newer blocks when an update fails) and start measuring how often that has issues instead.

In general, I hate retry logic as it hides more problematic underlying issues... I'm not 100% convinced this is the right approach to go.

### Test Plan

Added a unit test.

### CODE REVIEW

File: crates/shared/src/maintenance.rs
Comment (MartinquaXD) [inline crates/shared/src/maintenance.rs:82]: Should this maybe also log that the maintenance failed? Otherwise logs might be confusing if you see multiple updates on the same block without any indication of a problem.

File: crates/shared/src/maintenance.rs
Comment (MartinquaXD) [inline crates/shared/src/maintenance.rs:190]: One thing I still find weird is that if a component expects to run the maintenance on a new block it would make sense to pass that block into the maintenance function. Then the test could have asserted that you got the same block twice in a row.
I assume the maintenance function doesn't expect the block to get passed in to be usable even if updates are not triggered based on block updates.

**Review by MartinquaXD — APPROVED** [review]
> In general, I hate retry logic as it hides more problematic underlying issues... I'm not 100% convinced this is the right approach to go.

I think it's fine in this case because you are adding metrics so we should still be able to see that something is wrong.

**Comment by vkgnosis** [comment]
> This is done in order to reduce alert noise because of maintenance update delays.

How often does this happen? The obvious knob to turn for this is to make the alert condition more lenient. Why not do that? Whether we should retry here should depend on whether it makes sense for this functionality, not the alert.

**Review by sunce86 — APPROVED** [review]
Lg. 
After merging we can drop https://github.com/cowprotocol/services/pull/590

File: crates/shared/src/maintenance.rs
Comment (sunce86) [inline crates/shared/src/maintenance.rs:82]: One cool thing would be to include the maintainer name in metrics in case of failure.

**Comment by nlordell** [comment]
> Whether we should retry here should depend on whether it makes sense for this functionality, not the alert.

So, this is the reason I'm not 100% convinced this is the way to go (as specified in the description).

Specifically, if we want maintenance to run so we are at the "latest" block always, then retrying makes sense. Otherwise, we stay 12 seconds knowing that there is a new block and we have stale data before trying to update again. So, conceptually, retrying here is not so much of a sin.

However, my worry is that we are hiding node issues with this too much.

> Why not do that?

The issue is that the current alert measures how stale our data is (i.e. how far the last updated block is from the latest block). This PR proposes actively trying to make our data less stale (retrying to update to newer blocks when an update fails) and start measuring how often that has issues instead.

Does this make sense?

File: crates/shared/src/maintenance.rs
Comment (nlordell) [reply crates/shared/src/maintenance.rs:190]: We don't pass in blocks to the maintenance runners. True that this would make sense... however I wouldn't do it in this PR.

File: crates/shared/src/maintenance.rs
Comment (nlordell) [reply crates/shared/src/maintenance.rs:82]: Agreed. Unfortunately we don't have maintainer names yet. This is something that we could do in a separate PR.

Comment (nlordell) [reply crates/shared/src/maintenance.rs:85]: https://github.com/cowprotocol/services/issues/819

**Comment by vkgnosis** [comment]
Yeah makes sense. In the description it sounded like the only reason for this change was to reduce alert noise.

**Review by vkgnosis — APPROVED** [review]

File: crates/shared/src/maintenance.rs
Comment (nlordell) [reply crates/shared/src/maintenance.rs:193]: I think we just need a trait for 1 (update on new blocks).

For the other trait, they can just spawn a Tokio task.

**Comment by nlordell** [comment]
> In the description it sounded like the only reason for this change was to reduce alert noise.

Will update the PR description.

**Comment by MartinquaXD** [comment]
Just noticed that the `ServiceMaintenance` can maintain multiple different update tasks. If only a single task in the list failed to update all of the successful tasks would also get triggered again. I guess it would make sense to adapt the retry logic to only update the failed tasks until all tasks ran successfully for the given block.

**Comment by nlordell** [comment]
> I guess it would make sense to adapt the retry logic to only update the failed tasks until all tasks ran successfully for the given block.

In theory, their `update`s should be no-ops (at least its the case for the event updaters).

**Pull request merged**
